src/services/GithubService.ts

```diff
@@ -0,0 +1,106 @@
+import IRestClient, {IRestResponse} from "../rest/RestClient";
+import URLBuilder from "./URLBuilder";
+import {CommitInfo} from "../interfaces/GitHubTypes";
+import ResponseParser from "./ResponseParser";
+import {ICommitCache} from "./GitCommitCache";
+
+export interface GithubServiceError extends Error {
+    message: string,
+}
+
+export default class GithubService {
+
+    private readonly restClient: IRestClient;
+    private readonly urlBuilder: URLBuilder;
+    private readonly responseParser: ResponseParser;
+    private readonly cache: ICommitCache;
+
+    constructor(restClient: IRestClient, cache: ICommitCache) {
+        this.restClient = restClient;
+        this.urlBuilder = new URLBuilder( "3fc9ac24e39a5705954fadf7c30633ae9193c007");
+        this.responseParser = new ResponseParser();
+        this.cache = cache;
+    }
+
+    public async getAndSaveAllCommits(repoUrl: string): Promise<Array<CommitInfo>> {
+        try {
+            const historyExists: boolean = await this.cache.exists(repoUrl);
+            if (!historyExists) {
+                let totalNumCommits = await this.getNumCommits(repoUrl);
+                let retrievedCommits: Array<CommitInfo> = new Array<CommitInfo>();
+                while (retrievedCommits.length < totalNumCommits) {
+                    const commits = await this.getMoreCommits(retrievedCommits, repoUrl);
+                    retrievedCommits = retrievedCommits.concat(commits);
+                }
+                await this.cache.persistCommits(repoUrl, retrievedCommits);
+                return retrievedCommits;
+            }
+            return await this.cache.getCommitData(repoUrl);
+        } catch (err) {
+            console.warn(err);
+            throw { message: err.message } as GithubServiceError
+        }
+    }
+
+    private async getMoreCommits(retrievedCommits: Array<CommitInfo>, repoUrl: string): Promise<Array<CommitInfo>> {
+        let dateUpTo: string;
+        if (retrievedCommits.length === 0) {
+            dateUpTo = new Date().toISOString();
+        } else {
+            dateUpTo = retrievedCommits[retrievedCommits.length - 1].date;
```

Comment:
Need to subtract some time from `dateUpTo` or it will get that commit twice. Thats why the test was giving back 160 instead of 159.

Something like this if we want to stick with vanilla js dates:
```TypeScript
const lastRetrievedDate = retrievedCommits[retrievedCommits.length - 1].date;
dateUpTo = new Date(new Date(lastRetrievedDate).getTime() - 60000).toISOString();
```
or we can use a library like: https://date-fns.org/

Reply:
I’ll try to do it with vanilla JS before I try a library - don’t think we’ll work too much with dates anyway.